packages/nodes-base/nodes/Google/Gmail/Gmail.node.ts

```diff
@@ -0,0 +1,386 @@
+import {
+	IExecuteFunctions,
+} from 'n8n-core';
+
+import {
+	IDataObject,
+	INodeExecutionData,
+	INodeType,
+	INodeTypeDescription,
+} from 'n8n-workflow';
+
+import {
+	googleApiRequest,
+	encodeEmail,
+} from './GenericFunctions';
+
+import {
+	messageOperations,
+	messageFields,
+} from './MessageDescription';
+
+import {
+	labelOperations,
+	labelFields,
+} from './LabelDescription';
+
+import {
+	draftOperations,
+	draftFields,
+} from './DraftDescription';
+
+import {
+	isEmpty,
+} from 'lodash';
+
+export interface IEmail {
+	to?: string;
+	cc?: string;
+	bcc?: string;
+	inReplyTo?: string;
+	reference?: string;
+	subject: string;
+	body: string;
+	attachments?: IDataObject[];
+}
+
+interface IAttachments {
+	type: string;
+	name: string;
+	content: string;
+}
+
+export class Gmail implements INodeType {
+	description: INodeTypeDescription = {
+		displayName: 'Gmail',
+		name: 'gmail',
+		icon: 'file:gmail.png',
+		group: ['transform'],
+		version: 1,
+		subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
+		description: 'Consume the Gmail API',
+		defaults: {
+			name: 'Gmail',
+			color: '#d93025',
+		},
+		inputs: ['main'],
+		outputs: ['main'],
+		credentials: [
+			{
+				name: 'gmailOAuth2',
+				required: true,
+			},
+		],
+		properties: [
+			{
+				displayName: 'Resource',
+				name: 'resource',
+				type: 'options',
+				options: [
+					{
+						name: 'Draft',
+						value: 'draft',
+					},
+					{
+						name: 'Label',
+						value: 'label',
+					},
+					{
+						name: 'Message',
+						value: 'message',
+					},
+				],
+				default: 'draft',
+				description: 'The resource to operate on.',
+			},
+			//-------------------------------
+			// Draft Operations
+			//-------------------------------
+			...draftOperations,
+			...draftFields,
+			//-------------------------------
+			// Label Operations
+			//-------------------------------
+			...labelOperations,
+			...labelFields,
+			//-------------------------------
+			// Message Operations
+			//-------------------------------
+			...messageOperations,
+			...messageFields,
+		]
+	};
+
+	async execute(this: IExecuteFunctions): Promise<INodeExecutionData[][]> {
+		const items = this.getInputData();
+		const returnData: IDataObject[] = [];
+
+		const resource = this.getNodeParameter('resource', 0) as string;
+		const operation = this.getNodeParameter('operation', 0) as string;
+
+		let method = '';
+		let body: IDataObject = {};
+		let qs: IDataObject = {};
```

Comment:
I totally agree and was implementing it that way but learned that if a property is nested within another then TS does not allow you to set it. `body.message.raw` causes an `Object is possibly 'null' or 'undefined'` error. 😞

Reply:
It might be worth checking if the [optional chaining operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) fixes that.

Reply:
Oh wow I did not know about this!

Reply:
Unfortunately that gives me this error `The left-hand side of an assignment expression may not be an optional property access.`